### ISSUE TITLE
show user number

### ISSUE THREAD

**Issue by vieridipaola** [InnovateAsterisk/Browser-Phone#402]
Hi,

I'd like to display the user's "extension number". To do so I change this in phone.js:

in InitUI()

   $("#UserCallID").html(profileUser + " - " + profileName);

Maybe another custom setting via phoneOptions would be more appropriate?

**Comment by InnovateAsterisk** [comment]
To be clear ```profileUser``` as a property of XMPP settings and is the username associated with the XMPP account. So not everyone is using this feature.

DID's like 100, 101, etc don't really have anything to do with the inner workings of SIP, and are mostly applied via scripts etc on the Asterisk side. Typically one would have a ```username@domain.com``` structure, and a numerical DID as an alias making it easier to dial someone as 100 rather than ```vn38gks83d-sucfhvu3ckv9@domain.com``` in some configurations people use the did as the username eg ```100@domain.com``` and an alias of 100, this makes the relationship between the username and the DID difficult to change, but also assumes that the persons username is their DID. This may not always be the case, and im a fan of de-coupling these.

This reference describes the concept: http://www.asteriskdocs.org/en/3rd_Edition/asterisk-book-html-chunk/DeviceConfig_id283201.html  but again, im also not comfortable with MAC address as it has no place in WebRTC. 

Personally, I prefer a unique code like ```vn38gks83d-sucfhvu3ckv9@domain.com``` this makes it more secure, in that if someone was using DIDs as usernames, if I was to attempt a login, I already have half the information I require... this is something to avoid in a hosted situation!

There is a field that is available to use... It is part of the ```vCard```. Each User profile builds a vCard and should you enable XMPP it simply uploads this, if not then it keeps it locally. in the vCard, there are two free to use Number fields. I would use one of them.

And example of usage:
```
var profileVcard = getDbItem("profileVcard", null);
if(profileVcard != null) {
    profileVcard = JSON.parse(profileVcard);
    var DID = profileVcard.Number1;
}
```
👆🏼 I would be happy to look at implementing the above code.

This can be set in **Configure Extension** > **Appearance**, and also provisioned with something like: 
```
window.localStorage.setItem("profileVcard", JSON.stringify({
      TitleDesc: "<?php echo($Desc)) ?>",
      Mobile: "<?php echo($Mobile)) ?>",
      Email : "<?php echo($Email)) ?>",
      Number1 : "<?php echo($LocalDid)) ?>",
      Number2 : "<?php echo($InterBranchDid)) ?>",
}));
```

**Comment by vieridipaola** [comment]
Yes, sorry, I wasn't being clear. I want to display the DID so that the user knows how others can contact him/her. It might sound odd, but in a large organization with lots of people moving around, some might not know their own DID. Users authenticate via LDAP and their sip/xmpp credentials are auto-provisioned.  The SIP user part (phoneOptions.SipUsername) is not the same as their DID.

As a side question, when using XMPP would you recommend setting phoneOptions.profileUser = phoneOptions.SipUsername or would you set something else?

Using what you suggested sounds neat:

```
window.localStorage.setItem("profileVcard", JSON.stringify({
      TitleDesc: "<?php echo($Desc)) ?>",
      Mobile: "<?php echo($Mobile)) ?>",
      Email : "<?php echo($Email)) ?>",
      Number1 : "<?php echo($LocalDid)) ?>",
      Number2 : "<?php echo($InterBranchDid)) ?>",
}));
```

because in fact I would also like to show more than just the local DID. I might need to show the masked DID, the "real" local DID, and also an "external"/public DID if available (Number3?).
That's a lot to show on just one line so maybe #UserCallID should be left as it is with just the value of profileName, but made clickable to show the Configure Extension -> Appearance section in just one go.

**Comment by vieridipaola** [comment]
BTW, are the vCard phone number field names customizable?

Currently they are:

```
Mobile Number
Contact Number 1
Contact Number 2
```

I'd rather have something like:

```
External (public) number:
Internal (advertised) number:
Private internal number:
DECT number:
Pager:
Corporate mobile number:
Personal nuymber:
```

I might be asking too much...

**Comment by InnovateAsterisk** [comment]
> As a side question, when using XMPP would you recommend setting phoneOptions.profileUser = phoneOptions.SipUsername or would you set something else?

I would imagine that when people set up and Asterisk and OpenFire Server, the easiest thing to do is have the same Sip Username as Xmpp Username. At one point I think it was hard-coded like that, but for flexibility, you are welcome to configure as you like. ```phoneOptions.profileUser = phoneOptions.SipUsername``` seems like the obvious choice. 

The only thing worth noting here is that XMPP servers (mostly I assume) don't support multiple domains. So if you run an Asterisk PBX, for multiple companies, you cannot assign the XMPP server multiple domains. This is why the domain originally was hardcode to the sip domain, but now it will be set-able so XmppDomain is coming back in the next release. 

Also if you do want to try simulate a multi-domain setup, you can use a XmppRealm. This prepends a string and a XmppRealmSeparator to the username so ```100@domain.com``` can be  ```realm~100@commondomain.com```. This just makes the filtering of data according to the realm possible.
(Disclaimer: I didn't actually try this out much)

> because in fact I would also like to show more than just the local DID. I might need to show the masked DID, the "real" local DID, and also an "external"/public DID if available (Number3?). 

This is the idea, but there is limited space at the top there, especially when a buddy is selected. I was considering a clickable option at the top here, its always seemed like an obvious choice to try click/tab it... let me see what I can do.

**Comment by vieridipaola** [comment]
> but there is limited space at the top there, especially when a buddy is selected. I was considering a clickable option at the top here, its always seemed like an obvious choice to try click/tab it

This is enough for me:

```
$("#UserCallID").html(profileName);
$("#UserCallID").on('click', function(event){
	ShowMyProfile();
});
```

I would say that the only thing I'd change is the way to customize the vCard settings as in:
https://github.com/InnovateAsterisk/Browser-Phone/issues/402#issuecomment-1378477902

**Comment by InnovateAsterisk** [comment]
> BTW, are the vCard phone number field names customizable?

So XMPP use this: https://www.rfc-editor.org/rfc/rfc6351
Its an XML representation of the old vCard (https://www.rfc-editor.org/rfc/rfc6350)... yes it vomit inducing 🤮

The XML version is okay, but basically its about protocol now, so if you share your vCard, it has to be as vCard XML format, otherwise its rejected.

You will see I have selected just a few fields from an array of fields that you CAN submit. So to make it work on this UI, it has to establish a set of known nodes.

Also tho remember, its a shared vCard, so the way XMPP works, is that you can "exchange" vCards... meaning what you have populated in your vCard is whats sent to your buddy. It's already doing this to display Name, and picture of the buddy.

You can try change ```XmppSetMyVcard()``` to include more fields, and then ```XmppGetBuddyVcard(buddyObj)``` to read the buddy vCard. There you can also see how its currently working. You can probably get away with adding fields, but if you remove them, something may break.

**Comment by vieridipaola** [comment]
> You can try change `XmppSetMyVcard()` to include more fields, and then `XmppGetBuddyVcard(buddyObj)` to read the buddy vCard

Sure, I could add some fields there, and it should work. However, what about the Browser Phone UI in Configure Extension -> Appearance ? Maybe the UI elements shown there should be built from an array which can be customized in a "pre JS script". Or maybe if the admin user provides/defines an array, Browser Phone could load it and append it the the standard elements and data.

BTW Number2 is supposed to be a FAX number, but it's labeled as contact number 2 in the UI. Not that it's important or anything.
Anyway, making phone.js dynamically generate the UI contact elements and the vCard data from a customizable array would be nice.

For example, I would provide user data on page load via PHP (phone numbers of all types, title/FullName, e-mail) in an array for phone.js to use to:

1) populate vCard data
2) show UI elements approriately in Browser Phone -> Configure Extension -> Appearance

Is this feasible?
